### ISSUE TITLE
RareBytesTwo prefilter could cause false negative matches

### ISSUE THREAD

**Issue by quark-zju** [BurntSushi/aho-corasick#53]
Hi,

We recently got a user report about a path matching correctness issue in our library using `globset`. I have bisected the issue to https://github.com/BurntSushi/aho-corasick/commit/063ca0d253b1cfcef9f6b5533bee0d1fd2737c33. The reproduce test case can be found at https://gist.github.com/quark-zju/600f9f869f83384b87c66a4408c78bef#file-lib-rs-L523. I wonder if you have any quick insights about what went wrong in this area. Meanwhile I'm still trying to understand the related code and further minimize the test.

**Comment by BurntSushi** [comment]
If you can get me a reproduction using just globset, then I should be able to take it from there.

**Comment by BurntSushi** [comment]
Note that I can't seem to reproduce from your gist:

```
$ git clone https://gist.github.com/600f9f869f83384b87c66a4408c78bef.git
$ cd 600f9f869f83384b87c66a4408c78bef
$ cargo test
    Finished test [unoptimized + debuginfo] target(s) in 0.01s
     Running target/debug/deps/pathmatcher-6abf659345f9b625

running 15 tests
test tests::test_empty_negative ... ok
test tests::test_always_matcher ... ok
test tests::test_literal_negative ... ok
test tests::test_literal_paths ... ok
test tests::test_next_path_separator ... ok
test tests::test_negative_override ... ok
test tests::test_mixed_literal_and_complex_glob ... ok
test tests::test_mixed_literal_and_simple_glob ... ok
test tests::test_simple_glob ... ok
test tests::test_mixed_negative_literal_simple_glob ... ok
test tests::test_non_recursive_patterns ... ok
test tests::test_fast_paths ... ok
test tests::test_never_matcher ... ok
test tests::test_t62872044 ... ok
test tests::test_complex_glob ... ok

test result: ok. 15 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests pathmatcher

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

**Comment by quark-zju** [comment]
The gist has pinned globset to 0.4.2. Changing it to 0.4.3 reproduces the issue.

I have minimized the test case to:

```
#[test]
fn test_t62872044() {
    let mut builder = globset::GlobSetBuilder::new();
    let pats = ["aaandr/l/aaj/*", "xplat/hermes/*"];
    for pat in pats.iter() {
        builder.add(globset::GlobBuilder::new(pat).build().unwrap());
    }
    let set = builder.build().unwrap();
    assert!(!set.matches("aaandr/l/aaj/x").is_empty());
}
```

**Comment by BurntSushi** [comment]
Thanks! Minimized to `aho-corasick` usage:

```
    let ac = AhoCorasick::new_auto_configured(&["ab/j/", "x/"]);
    assert!(ac.is_match("ab/j/"));
```

**Comment by BurntSushi** [comment]
Fix up in #54. Thanks for the easy repro! Quite a gnarly bug

**Comment by BurntSushi** [comment]
Fix published in `aho-corasick 0.7.9` on crates.io.
